File: opencv-projects-python/kalman-tracking-predicting.py
# Import Necessary Libraries

import cv2
import numpy as np 
import matplotlib.pyplot as plt
import time


# Path to video  
video_path = r"videos/helicopter3.mp4"
video = cv2.VideoCapture(video_path)

# read only the first frame for drawing a rectangle for the desired object
ret,frame = video.read()

# I am giving  big random numbers for x_min and y_min because if you initialize them as zeros whatever coordinate you go minimum will be zero 
x_min,y_min,x_max,y_max=36000,36000,0,0


def coordinat_chooser(event,x,y,flags,param):
    global go , x_min , y_min, x_max , y_max

    # when you click the right button, it will provide coordinates for variables
    if event==cv2.EVENT_LBUTTONDOWN:
        
        # if current coordinate of x lower than the x_min it will be new x_min , same rules apply for y_min 
        x_min=min(x,x_min) 
        y_min=min(y,y_min)

         # if current coordinate of x higher than the x_max it will be new x_max , same rules apply for y_max
        x_max=max(x,x_max)
        y_max=max(y,y_max)

        # draw rectangle
        cv2.rectangle(frame,(x_min,y_min),(x_max,y_max),(0,255,0),1)


    """
        if you didn't like your rectangle (maybe if you made some misscliks),  reset the coordinates with the middle button of your mouse
        if you press the middle button of your mouse coordinates will reset and you can give a new 2-point pair for your rectangle
    """
    if event==cv2.EVENT_MBUTTONDOWN:
        print("reset coordinate  data")
        x_min,y_min,x_max,y_max=36000,36000,0,0

cv2.namedWindow('coordinate_screen')
# Set mouse handler for the specified window, in this case, "coordinate_screen" window
cv2.setMouseCallback('coordinate_screen',coordinat_chooser)


while True:
    cv2.imshow("coordinate_screen",frame) # show only first frame 
    
    k = cv2.waitKey(5) & 0xFF # after drawing rectangle press ESC   
    if k == 27:
        cv2.destroyAllWindows()
        break


# take region of interest ( take inside of rectangle )
roi_image=frame[y_min+2:y_max-2,x_min+2:x_max-2]
roi_rgb=cv2.cvtColor(roi_image,cv2.COLOR_BGR2RGB)

# convert roi to grayscale, SIFT Algorithm works with grayscale images
roi_gray=cv2.cvtColor(roi_image,cv2.COLOR_BGR2GRAY) 

# Initialize the FAST detector and BRIEF descriptor extractor
fast = cv2.FastFeatureDetector_create(threshold=1)
brief = cv2.xfeatures2d.BriefDescriptorExtractor_create()


# detect keypoints
keypoints_1 = fast.detect(roi_gray, None)
# descriptors
keypoints_1, descriptors_1 = brief.compute(roi_gray, keypoints_1)

# draw keypoints for visualizing
keypoints_image = cv2.drawKeypoints(roi_rgb, keypoints_1, outImage=None, color=(23, 255, 10))


# matcher object
bf = cv2.BFMatcher()

def detect_target_fast(frame):
    # convert frame to gray scale 
    frame_gray = cv2.cvtColor(frame, cv2.COLOR_BGR2GRAY)

    # Detect keypoints using FAST
    keypoints_2 = fast.detect(frame_gray, None)

    # Compute descriptors using BRIEF
    keypoints_2, descriptors_2 = brief.compute(frame_gray, keypoints_2)

    avg_x,avg_y=0,0

    """
    Compare the keypoints/descriptors extracted from the 
    first frame (from target object) with those extracted from the current frame.
    """
    if descriptors_2 is not None:
        matches = bf.match(descriptors_1, descriptors_2)
        
        if matches:
            # Initialize sums for x and y coordinates
            sum_x = 0
            sum_y = 0
            match_count = 0
            
            for match in matches:
                # .trainIdx gives keypoint index from current frame 
                train_idx = match.trainIdx
                
                # current frame keypoints coordinates
                pt2 = keypoints_2[train_idx].pt
                
                # Sum the x and y coordinates
                sum_x += pt2[0]
                sum_y += pt2[1]
                match_count += 1
            
            # Calculate average of the x and y coordinates
            avg_x = sum_x / match_count
            avg_y = sum_y / match_count
            
    return int(avg_x),int(avg_y)


# Initialize Kalman filter parameters
kalman = cv2.KalmanFilter(4, 2)   
 
kalman.measurementMatrix = np.array([[1, 0, 0, 0], [0, 1, 0, 0]], np.float32)
kalman.transitionMatrix = np.array([[1, 0, 1, 0], [0, 1, 0, 1], [0, 0, 1, 0], [0, 0, 0, 1]], np.float32)
kalman.processNoiseCov = np.eye(4, dtype=np.float32) * 0.03  # Process noise
kalman.measurementNoiseCov = np.eye(2, dtype=np.float32) * 0.5  # Measurement noise


# Startcapturing the video from file
cap = cv2.VideoCapture(video_path)

while True:
    ret, frame = cap.read()
    if not ret:
        break
    
    # Predict the new position of the ball
    predicted = kalman.predict()
    predicted_x, predicted_y = int(predicted[0]), int(predicted[1])
    predicted_dx, predicted_dy = predicted[2], predicted[3]  # Predicted velocity

    print(predicted_x, predicted_y )
    print(f"Predicted velocity: (dx: {predicted_dx}, dy: {predicted_dy})")

    
    # Detect the ball in the current frame
    ball_position = detect_target_fast(frame)
    
    if ball_position:
        measured_x, measured_y = ball_position
        # Correct the Kalman Filter with the actual measurement
        kalman.correct(np.array([[np.float32(measured_x)], [np.float32(measured_y)]]))
        # Draw the detected ball
        cv2.circle(frame, (measured_x, measured_y), 6, (0, 255, 0), 2) # green --> correct position
    
    # Draw the predicted position (Kalman Filter result)
    cv2.circle(frame, (predicted_x, predicted_y), 8, (0, 0, 255), 2) # red --> predicted position

    # Show the frame
    cv2.imshow("Kalman Ball Tracking", frame)
    
    # Break on 'q' key press
    if cv2.waitKey(30) & 0xFF == ord('q'):  # 30 ms delay for smooth playback
        break

cap.release()
cv2.destroyAllWindows()

File: opencv-projects-c++/ObjectTracking-kalman/kalman-tracking-predicting.cpp
#include <opencv2/opencv.hpp>
#include <opencv2/xfeatures2d.hpp>
#include <iostream>

using namespace cv;
using namespace std;

// Global variables for rectangle coordinates
int x_min = 36000, y_min = 36000, x_max = 0, y_max = 0;
Mat frame;

// Function to handle mouse events for selecting ROI
void coordinateChooser(int event, int x, int y, int flags, void* param) {
    if (event == EVENT_LBUTTONDOWN) {
        // Update min and max coordinates
        x_min = min(x, x_min);
        y_min = min(y, y_min);
        x_max = max(x, x_max);
        y_max = max(y, y_max);

        // Draw rectangle on the frame
        rectangle(frame, Point(x_min, y_min), Point(x_max, y_max), Scalar(0, 255, 0), 1);
    }

    if (event == EVENT_MBUTTONDOWN) {
        // Reset coordinates if middle mouse button is pressed
        cout << "Reset coordinate data" << endl;
        x_min = 36000; y_min = 36000; x_max = 0; y_max = 0;
    }
}

// Function to detect target using FAST and BRIEF
Point detectTargetFast(Mat& frame, Ptr<FastFeatureDetector>& fast, Ptr<xfeatures2d::BriefDescriptorExtractor>& brief,
                      const vector<KeyPoint>& keypoints_1, const Mat& descriptors_1, BFMatcher& bf) {
    Mat frame_gray;
    cvtColor(frame, frame_gray, COLOR_BGR2GRAY);

    // Detect keypoints in the current frame
    vector<KeyPoint> keypoints_2;
    fast->detect(frame_gray, keypoints_2);

    // Compute descriptors
    Mat descriptors_2;
    brief->compute(frame_gray, keypoints_2, descriptors_2);

    int avg_x = 0, avg_y = 0;
    if (!descriptors_2.empty()) {
        vector<DMatch> matches;
        bf.match(descriptors_1, descriptors_2, matches);

        if (!matches.empty()) {
            int sum_x = 0, sum_y = 0, match_count = 0;

            for (const auto& match : matches) {
                // Get the coordinates of the matched keypoint
                Point2f pt2 = keypoints_2[match.trainIdx].pt;
                sum_x += pt2.x;
                sum_y += pt2.y;
                match_count++;
            }

            avg_x = sum_x / match_count;
            avg_y = sum_y / match_count;
        }
    }
    return Point(avg_x, avg_y);
}

int main() {
    // Load the video
    string video_path = "/home/omer/tracking-projects/opencv-projects-c++/videos/plane (1).mp4";
    VideoCapture video(video_path);

    if (!video.isOpened()) {
        cerr << "Error: Could not open video." << endl;
        return -1;
    }

    // Read the first frame
    video >> frame;
    if (frame.empty()) {
        cerr << "Error: Could not read frame." << endl;
        return -1;
    }

    // Set up mouse callback
    namedWindow("coordinate_screen", 1);
    setMouseCallback("coordinate_screen", coordinateChooser);

    while (true) {
        imshow("coordinate_screen", frame);
        char k = waitKey(5);
        if (k == 27) { // Press ESC to break
            destroyAllWindows();
            break;
        }
    }

    // Extract ROI from the selected rectangle
    Mat roi_image = frame(Rect(x_min + 2, y_min + 2, x_max - x_min - 4, y_max - y_min - 4));
    Mat roi_gray;
    cvtColor(roi_image, roi_gray, COLOR_BGR2GRAY);

    // Initialize FAST and BRIEF
    Ptr<FastFeatureDetector> fast = FastFeatureDetector::create(1);
    Ptr<xfeatures2d::BriefDescriptorExtractor> brief = xfeatures2d::BriefDescriptorExtractor::create();

    // Detect keypoints and compute descriptors for the ROI
    vector<KeyPoint> keypoints_1;
    fast->detect(roi_gray, keypoints_1);
    Mat descriptors_1;
    brief->compute(roi_gray, keypoints_1, descriptors_1);

    // Initialize Kalman filter
    KalmanFilter kalman(4, 2, 0);
    kalman.measurementMatrix = (Mat_<float>(2, 4) << 1, 0, 0, 0, 0, 1, 0, 0);
    kalman.transitionMatrix = (Mat_<float>(4, 4) << 1, 0, 1, 0, 0, 1, 0, 1, 0, 0, 1, 0, 0, 0, 0, 1);
    setIdentity(kalman.processNoiseCov, Scalar::all(0.03));
    setIdentity(kalman.measurementNoiseCov, Scalar::all(0.5));

    // Start processing the video
    VideoCapture cap(video_path);
    if (!cap.isOpened()) {
        cerr << "Error: Could not open video." << endl;
        return -1;
    }

    BFMatcher bf(NORM_HAMMING);

    while (true) {
        Mat frame;
        cap >> frame;
        if (frame.empty()) {
            break;
        }

        // Predict the new position using Kalman filter
        Mat prediction = kalman.predict();
        Point predicted_pt(prediction.at<float>(0), prediction.at<float>(1));

        // Detect the target in the current frame
        Point measured_pt = detectTargetFast(frame, fast, brief, keypoints_1, descriptors_1, bf);

        if (measured_pt.x != 0 && measured_pt.y != 0) {
            // Correct the Kalman filter with the measurement
            Mat_<float> measurement(2, 1);
            measurement(0) = measured_pt.x;
            measurement(1) = measured_pt.y;
            kalman.correct(measurement);

            // Draw the measured position
            circle(frame, measured_pt, 6, Scalar(0, 255, 0), 2); // Green circle for the measured position
        }

        // Draw the predicted position
        circle(frame, predicted_pt, 8, Scalar(0, 0, 255), 2); // Red circle for the predicted position

        // Display the frame
        imshow("Kalman Ball Tracking", frame);

        // Break on 'q' key press
        if (waitKey(30) == 'q') {
            break;
        }
    }

    cap.release();
    destroyAllWindows();
    return 0;
}
